pkg/prometheus/promcfg.go

```diff
@@ -1142,6 +1143,263 @@ func (cg *ConfigGenerator) generatePodMonitorConfig(
 	return cfg
 }
 
+func (cg *ConfigGenerator) generatePodMonitorConfigForAgentDaemonSet(
+	m *monitoringv1.PodMonitor,
+	ep monitoringv1.PodMetricsEndpoint,
+	i int, apiserverConfig *monitoringv1.APIServerConfig,
+	store *assets.StoreBuilder,
+) yaml.MapSlice {
+	scrapeClass := cg.getScrapeClassOrDefault(m.Spec.ScrapeClassName)
+
+	cfg := yaml.MapSlice{
+		{
+			Key:   "job_name",
+			Value: fmt.Sprintf("podMonitor/%s/%s/%d", m.Namespace, m.Name, i),
+		},
+	}
+	cfg = cg.AddHonorLabels(cfg, ep.HonorLabels)
+	cfg = cg.AddHonorTimestamps(cfg, ep.HonorTimestamps)
+	cfg = cg.AddTrackTimestampsStaleness(cfg, ep.TrackTimestampsStaleness)
+
+	attachMetaConfig := mergeAttachMetadataWithScrapeClass(m.Spec.AttachMetadata, scrapeClass, "2.35.0")
+
+	s := store.ForNamespace(m.Namespace)
+
+	k8SSDConfig := cg.generateK8SSDConfig(m.Spec.NamespaceSelector, m.Namespace, apiserverConfig, s, kubernetesSDRolePod, attachMetaConfig)
+	k8SSDConfigValue := k8SSDConfig.Value.([]yaml.MapSlice)
```

Comment:
```suggestion
k8SSDConfig.Value = append(k8SSDConfig.Value.([]yaml.MapSlice), yaml.MapSlice{
	{
		Key:   "role",
		Value: "pod",
	},
	{
		Key: "selectors",
		Value: yaml.MapSlice{
			{
				Key:   "role",
				Value: "pod",
			},
			{
				Key:   "field",
				Value: "spec.nodeName=" + os.Getenv("NODE_NAME"),
			},
		},
	},
})
```

I'm not sure if it works, but wdyt?

Reply:
Basically it's combining my two steps into one, so it'll probably work. Let me see